Add adjustable weapon volume

swamproyal opened the pull request:
This allows you to set weapon related sound volumes via a convar float. This is a pretty popular request on the server where you like to hear player sounds and you're in a theater or whatever and the weapons get annoying. Could probably also be added to the tab menu settings.

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Update README.md

swampservers committed:
Update readme_format.md

LegacyValken committed:
Adjustments to dev tools. (#388)

* Update ponyspeak.lua

* Revert taunt fix.

* Update cl_devtools.lua

New player meta function IsLocalDev
New dev function model

* Update cl_devtools.lua

* Spacing

* Update cl_devtools.lua

Metroid55 committed:
Changes/Fixes (#387)

* Changes/Fixes

-fixed golfclub description
-fixed gauntlet description
-fixed snap title
-added purpose text to anonymous mask
-added more words to ponify and hopefully fixed all uppercase for non ponified words
-added more emotes to owoifier
-added 'givepointsrandom' chat command (please review if I did this correctly, because I don't have access to the givepoints lua file)

* update randomgivepoints

* ponify optimize

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Merge branch 'master' of github.com:swampservers/contrib

swampservers committed:
Garry's Mod

swampservers committed:
Lint

Metroid55 committed:
fix (#391)

This one broke the owoifier

swampservers committed:
stuff

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

Commit:
BFlix Fix (#393)

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Add pCasino/Floppa NPC content

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

Metroid55 committed:
Button Chat Fixes (#394)

* Button Chat Fixes

Should fix messages getting cut off and made them more coherent

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Add Kefta to CODEOWNERS

BananaInk committed:
Updated lighting props (#397)

swampservers committed:
Lint

swampservers committed:
Fix Lint/missing commas in PropTrashSpecialModels

swampservers committed:
Move AniMixPlay to defunct; it's dead

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

WinterPhoenix committed:
Restore cl_crashdetect.lua to Contrib

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

horatio42 committed:
Update sh_neko.lua (#399)

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swamproyal committed:
Add royal to devtools (#400)

swampservers committed:
Garry's Mod

WinterPhoenix committed:
Add YugenAnime service to Contrib

WinterPhoenix committed:
Add base cinema service to contrib

swampservers committed:
Garry's Mod

swampservers committed:
Merge branch 'master' of github.com:swampservers/contrib

swamproyal committed:
Add kick.com support (#402)

swampservers committed:
Lint

swampservers committed:
Merge branch 'master' of github.com:swampservers/contrib

swamproyal committed:
Important change (#404)

swamproyal committed:
Fix m3u URI parsing (#401)

swampservers committed:
Lint

swamproyal committed:
Add swamp_weaponvolume

swamproyal committed:
Formatting

swamproyal committed:
Formatting, add file header

Kefta commented on lua/cinema/misc/cl_weaponvolume.lua:
```diff
@@ -0,0 +1,23 @@
+}
+
+hook.Add("EntityEmitSound", "EntityEmitSound_WeaponVolume", function(data)
+    for k, v in pairs(data) do
```
Remove the loop and access the SoundName key directly like ``local sound = data.SoundName``

Kefta replied:
I see you beat me to it

Kefta commented on lua/cinema/misc/cl_weaponvolume.lua:
```diff
@@ -0,0 +1,23 @@
+-- This file is subject to copyright - contact swampservers@gmail.com for more information.
+local swamp_weaponvolume = CreateClientConVar("swamp_weaponvolume", "1", true, false, "", 0.0, 1.0)
+
+local weapon_sound_patterns = {
```
What are the exact file paths that you are trying to capture with these patterns? This will probably catch some false positives without a stricter pattern.

swamproyal replied:
Any sounds emitted by weapons which cause damage to the player. I'm not sure a stricter pattern is necessary, tbh.

Kefta replied:
Stricter patterns would prevent any possible chance of false positives passing through if any of these names are used in part of the directory or file name for non-weapons + be slightly more efficient with how patterns are implemented.

A better implementation might be to skip the patterns and instead just check if the entity emitting the sound is a weapon, making it future-proof as well. There might be cases in Swamp SWEPs where the sound is emitted from the player instead, but either a hack can be added to still check for the patterns + the player's active weapon, or I can go in and fix those cases.

swamproyal replied:
So I think checking if the entity emitting the sound is a weapon isn't so desirable because a lot of the SWEPs are weapons if I'm not mistaken - such as the "free toys". This feature should not affect those sounds. Like I mentioned, I think this should only affect weapons which cause damage to a player.

swamproyal replied:
Side note: I thought about branding this as "adjustable gun sounds" but I wanted it to capture more SWEPs than just guns, such as jihad bomb, throatneck slitter, big bomb, etc...

Kefta replied:
In that case, though it sounds like more complicated option, I think it would be best to insert something like SWEP.IsSwampDamageWeapon = true into all the SWEPs that should be affected, then the hook would check for that key in the entity that emitted the sound. This would prevent targeting any unintentional sounds, make it easily expandable for SWEP authors without having to remember to go into this file, and be far more efficient (not that this is currently a frame killer or anything). Guns should be easy to apply this property to since they all resolve to their own common weapon bases, but other than the weapons you just listed, what else should receive this property?

Separately, a convar could be added to change the volume of all weapon sounds regardless of their type but I don't know if you wanted to not affect those sounds for a reason.

swamproyal replied:
Probably a better option here is to effectively have a whitelist on the toys*. I think these are the only SWEPs which should remain unaffected by the weapon volume. It would be annoying to enumerate all of the weapons which need to be affected because that includes all of the CS stuff, HL2 stuff, a mix of custom swamp stuff, probably other stuff. Would you be content with me adding a property to the toys e.g. `SWEP.IsToy = true`?

~~To clarify, should we get the Entity from [EntityEmitSound](https://wiki.facepunch.com/gmod/GM:EntityEmitSound) table, get the active weapon, i.e. `Entity:GetActiveWeapon()`, then check if `IsToy` exists? If so, apply the volume scaling?~~

How do you recommend getting the entity which triggered the sound from the EntityEmitSound callback table?

*Specifically, the toys are these if you aren't familar: https://imgur.com/H37JZWW.png

Kefta replied:
Good idea, could you call it something like``SWEP.IsPassiveWeapon`` (specifically with the Weapon suffix since IsPassive could conflict with some weapon base's "IsPassive" method for checking weapon lowering) to notate that it should apply to anything that doesn't shoot, blow up, or otherwise do damage? I would imagine the Peacekeeper (and the taser and pickaxe?) should also be affected by this convar despite being a toy(s).

I forgot the weapons were public so you can change this if need-be, but the idea is that the entity should be used directly and no hacks would be needed. If any SWEP is running EmitSound on the player instead of the weapon, it needs to be corrected in code. I see that pretty much everything uses ExtEmitSound so that probably won't be a problem.

Kefta replied:
Actually, there are some valid cases in which a damaging weapon will create sounds on a non-weapon entity (taser, crusader's sword) and some incorrect ones that do need to be fixed (the garand). However, due to this, the solution of checking a weapon property alone won't work. I need to do some brainstorming but it might require solidifying usage of the ExtEmitSound function or overriding EmitSound in the metatable.

swamproyal replied:
>then the hook would check for that key in the entity that emitted the sound. 

How would this be possible?

EntityEmitSound callback data table contains an Entity property which represents the Entity that generated the sound. In the case of a gun shot sound, that entity is the Player. There isn't a way to know if the sound event was generated from a SWEP/Weapon. This would only work for actual weapon entities, e.g. bomb, mines, etc...

swamproyal replied:
Bump @Kefta, any thoughts on my comment? I'm not sure that we can only use the entity in EntityEmitSound callback data table without looking at the sound file.

Kefta replied:
Still thinking, I'm looking further to a swamp-specific EmitSound override that would replace both ExtEmitSound and provide more info to this hook

swamproyal replied:
Any further thoughts? People would like to start using this. We could just go back to my original implementation with the regex patterns and just make them more specific.

swamproyal committed:
Optimize table entry access

Kefta commented on lua/cinema/misc/cl_weaponvolume.lua:
```diff
@@ -0,0 +1,23 @@
+        if (k == "SoundName") then
+            for i = 1, #weapon_sound_patterns do
+                if string.find(v, weapon_sound_patterns[i]) then
+                    data.Volume = swamp_weaponvolume:GetFloat()
```
This should be scaled by the previously defined Volume (if there is one) in case the weapon specifies its own volume in EmitSound. ``data.Volume = swamp_weaponvolume:GetFloat() * (data.Volume or 1)``

swamproyal replied:
Thanks, good catch

swamproyal committed:
Scale weapon volume

swampservers committed:
Garry's Mod

swamproyal committed:
Fix starting volume for kick videos (#410)

* Kick player volume changes

swampservers committed:
Lint

Kefta committed:
Change some EmitSound targets from the player to the weapon (#407)

swampservers committed:
Lint

swamproyal committed:
Revert to using file.html for .m3u8 (#408)

swamproyal commented:
Bump @Kefta

swamproyal committed:
Remove broken/useless chat commands (#415)

swamproyal committed:
Bflix fix (#414)

swampservers committed:
Lint

swamproyal committed:
Add rumble.com support (#411)

swampservers committed:
Lint

swamproyal committed:
Fix bflix load after some time has passed since video queue (#417)

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swamproyal committed:
Fix rumble seek (#416)

swamproyal committed:
Fix lua error in weapon_crusadersword

Kefta committed:
Merge pull request #421 from swamproyal/crusadersword-author

Fix lua error in weapon_crusadersword

swamproyal committed:
Move oligopoly to defunct (#418)

swamproyal committed:
Optimize bflix (#419)

swamproyal committed:
Fix pony eye texture on Linux (#423)

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

Commit:
Fix YouTube sometimes having max volume on start, Appearance adjustments for BFlix (#424)

swampservers committed:
Garry's Mod

BananaInk committed:
Updated lighting props (#425)

swampservers committed:
Lint

swamproyal committed:
Add Infowars video service (#412)

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swamproyal committed:
Reenable snow by default in December (#427)

swampservers committed:
Lint

swamproyal committed:
Fix infowars url pattern match (#428)

Commit:
Fix fsd-overrun-toy.mdl physics noise pollution (#426)

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

Commit:
Magic Button and Outfitter Fixes (#429)

* fixed bounty functions in magic button

* remove redundant code

* outfitter fix

* .

* .

swampservers committed:
Lint

swamproyal committed:
Fix bad tag condition (#431)

swamproyal committed:
Update devtools structEnvironment paths (#409)

WinterPhoenix reviewed (commented):
So, I think this should be added to the Scoreboard UI before it's merged, along with the other volume sliders. Beyond that it's probably good to go...although EntityEmitSound is generally unreliable so YMMV

swampservers commented:
This should probably be implemented alongside the main clientside volume stuff to make sure it doesn't conflict.

Also: There are too many volume sliders. This and the loading volume should be in a collapsable/popout or something.

swamproyal committed:
Merge branch 'master' into weaponvolume

swamproyal commented:
> So, I think this should be added to the Scoreboard UI before it's merged, along with the other volume sliders. Beyond that it's probably good to go...although EntityEmitSound is generally unreliable so YMMV

I tested it on the server again and it works still. I don't think I have access to the scoreboard code so I can't help with that.

swamproyal committed:
Merge branch 'master' into weaponvolume

WinterPhoenix commented:
Of course it's in restricted for some reason...

Edit - It's here now: https://github.com/swampservers/contrib/blob/master/lua/cinema/cl_ui/settings.lua

WinterPhoenix committed:
Move cl_ui/settings.lua from restricted

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

Commit:
Casino decal fixes (#432)

swamproyal committed:
Improve thirdperson (#403)

* Revive #285 third person improvements

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Lint

WinterPhoenix committed:
Fix cl_calcview merge conflict

swampservers committed:
Garry's Mod

swamproyal committed:
Thirdperson cleanup (#433)

swampservers committed:
Lint

swampservers committed:
Garry's Mod

WinterPhoenix committed:
Add non-map assets for Map v4

swamproyal committed:
Fix drawing player when using selfie stick (#434)

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swamproyal committed:
Fix render crosshair in thirdperson (#435)

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Lint

WinterPhoenix committed:
Restoring weapon_doom3_base (wtf?)

WinterPhoenix committed:
- Entities: Add skeleton, ore, and wormhole to contrib
- Effects: Add ore chunks, skeleton splatter, and skeleton form to contrib
- Add church spotlight map effect to contrib

swampservers committed:
Garry's Mod

PYROTEKNIK committed:
included sparkle material (#437)

WinterPhoenix committed:
Add location damage code to contrib

WinterPhoenix committed:
Add our CNavArea:(Set/Get)Place override to contrib (used by Skeletons)

swampservers committed:
Garry's Mod

PYROTEKNIK committed:
Skeleton Spawn Rework + AI Improvements (#438)

Skeletons will now remain hidden where they spawn and pop out when a player approaches. Will re-hide if player leaves a larger radius.

PYROTEKNIK committed:
Maze Materials. I don't think these were packed to the map? (#439)

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Lint

swampservers committed:
Garry's Mod

PYROTEKNIK committed:
Fixed Missing Maze Column Texture (#440)

PYROTEKNIK committed:
Skeleton tweaks but they actually work this time (#441)

swampservers committed:
Lint

swampservers committed:
Garry's Mod

WinterPhoenix committed:
Add Memo and List types to contrib

swamproyal committed:
Update BFlix domain (#445)

PYROTEKNIK committed:
Ore spawning reworks, Diamond ore variant (#444)

Cleaned up spawning function, Removed a bunch of pointless code leftover from buttons
gave it vphysics collision
added diamond variant
changed mining title, added stat effect

swampservers committed:
Lint

PYROTEKNIK committed:
Skeleton Fix Again (#443)

Oops! looks like the spawning breaks somehow.
Here's a more secure spawn function.

use the global var i set for limiting

* added title for skeleton kills

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swamproyal committed:
Fix BFlix/Nites (#446)

swampservers committed:
Lint

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swampservers committed:
Garry's Mod

swamproyal committed:
BFlix/Nites Fixes (#448)

swamproyal committed:
Add fmoviesz video service (#449)

swamproyal committed:
Add WCO video service (#450)

swampservers committed:
Lint

swamproyal committed:
Merge branch 'master' into weaponvolume

swamproyal committed:
Add scoreboard slider for weapon volume

swamproyal commented:
@WinterPhoenix I added a scoreboard slider for the weapon volume. If you think that is too many sliders on the scoreboard menu, I think we should add this as a convar only setting regardless.
